Behemoth V2 - Complete Rework

lKiseki opened the pull request:
## About The Pull Request
### [Check out the design document here.](https://docs.google.com/document/d/1yl7zEJF0De7kiEEDbce-0Ii8GdX_iYIixUpzCs0TN48/edit?usp=sharing)
⠀
Completely reworks Behemoth from the ground up. Its role in a hive has been redesigned, and its abilities have been changed to better accommodate for what is intended and expected of it. Visual and sound effects have also been changed here and there in lieu of changes and additions. Lastly, this PR also refactors most of the coding related to this caste.

To do:
- **(OPTIONAL)** Replace Earth Pillar's icon with a basketball for _Silly Mode_.
- **(OPTIONAL)** Complete resprite of Behemoth (commissioned GregorMN for this).

## Why It's Good For The Game
Behemoth has been in a weird spot ever since its introduction into the game. It tried to do various things at once, and it failed to compete with marines and most other castes. This PR aims to properly define its identity and reflect that in its kit, while also making it an attractive pick for our players.

## Credits
- Visual effects taken from **Duelyst** (licensed under CC0).
- Sound effects taken from **Pixabay** (free to use).
  - Earth Pillar's landing sound is [Rock Cinematic](https://pixabay.com/sound-effects/rock-cinematic-161648/), made by [LordSonny](https://pixabay.com/users/lordsonny-38439655/).
  - Landslide's impact sound is [Cinematic Impact Hit](https://pixabay.com/sound-effects/cinematic-impact-hit-352702/), made by [Universfield](https://pixabay.com/users/universfield-28281460/).
  - Geocrush's hit sound is [Explosion FX](https://pixabay.com/sound-effects/explosion-fx-343683/), made by [SoundReality](https://pixabay.com/users/soundreality-31074404/).

## Changelog
:cl: Lewdcifer
add: Behemoth has been COMPLETELY reworked from the ground up.
spellcheck: The changelog only offers a very brief overview, so if you're interested in specifics, refer to the pull request.
add: Added a new ability, Seize. Fast movement option that requires Earth Pillars to function.
add: Added a new ability, Geocrush. Heavy melee damage.
del: Rolling ability removed.
del: Seismic Fracture removed.
balance: Redesigned Earth Riser, as well as Earth Pillars, to be faster, resilient, and versatile, but more limited in access, damage, and range.
balance: You can now climb and jump over Earth Pillars.
balance: Landslide redesigned to be much faster in execution, similarly to its now deprecated Primordial variant, at the expense of damage and reach.
balance: Primal Wrath, the Behemoth's primordial ability, has been completely redesigned. Instead of upgrading all of your abilities, it will now restore all damage taken in a time window, while providing a few seconds of invulnerability, in exchange for a stacking temporary debuff.
balance: Allied xenos can no longer ride the Behemoth.
add: Added various statistic trackers for several Behemoth actions.
soundadd: Added a few new sounds for Earth Pillars, Landslide, and Geocrush.
sounddel: Removed unused sounds related to Behemoth.
imageadd: Added some new sprites for Earth Pillars, Landslide, and Geocrush. Also reorganized Behemoth's stuff.
imagedel: Removed unused sprites related to Behemoth.
fix: Squashed some bugs and fixed some glitches or unintended behavior related to Behemoth.
refactor: Refactored most, if not all, coding related to Behemoth.
/:cl:

lKiseki committed:
initial commit for the behemoth rework. missing content

lKiseki committed:
Merge branch 'master' of https://github.com/tgstation/TerraGov-Marine-Corps into beherework

lKiseki committed:
behemoth rework: cleans up, finishes primordial ability

lKiseki committed:
behemoth rework: more clean up, balance pass

lKiseki committed:
behemoth rework: final changes, incl. balance and fixes

lKiseki committed:
behemoth rework: appease the integration tests

lKiseki committed:
behemoth rework: go away integration tests

lKiseki committed:
behemoth rework: leave me alone integration tests

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/castedatum_behemoth.dm:
```diff
@@ -10,35 +10,31 @@
 
 	// *** Melee Attacks *** //
-	melee_damage = 23
+	melee_damage = 30
```
I'm not really looking at the balance right now, but why is bemo doing better damage than a rav, while being more tanky, and being stager immune and still having all the funny abilities?

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+	playsound(loc, 'sound/effects/alien/behemoth/earth_pillar_destroyed.ogg', 40, TRUE)
+	new /obj/effect/temp_visual/behemoth/earth_pillar/creation/destruction(loc)
+	if(xeno_owner)
+		SEND_SIGNAL(src, COMSIG_EARTH_PILLAR_DESTROY) // Surprised Destroy() doesn't send a signal, but hey, we make do.
```
?? there is a signal, its qdeleting

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+	return ..()
 
-/obj/effect/temp_visual/behemoth/stomp/east/Initialize(mapload, direction)
+/* to do: need damage icons
```
hi todo

lKiseki replied:
Yeah, I do need sprites for Earth Pillar. I'll see if I can get those sometime soon.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+	/// The atom that's currently holding this object.
+	var/mob/living/carbon/xenomorph/current_holder
+	/// References the dummy item created when this object is held.
+	var/obj/item/pillar_item/dummy_item
```
hard deling?

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+/obj/structure/xeno/earth_pillar/proc/pre_throw(datum/source)
+	SIGNAL_HANDLER
+	climbable = FALSE
+	pixel_y = initial(pixel_y)
```
pixel x and y determine the actual turf you are on for various code and visual calculations, should this be using z and w?

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
-	plasma_use_multiplier = 0
-	agile_charge = TRUE
-	should_start_on = FALSE
+#define SEIZE_PASS_FLAGS (PASS_LOW_STRUCTURE|PASS_MOB|PASS_XENO|PASS_THROW|PASS_PROJECTILE)
```
giving them projectile and throw immunity is pretty wowza

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+				xeno_owner.balloon_alert(xeno_owner, "Wait [cooldown_remaining()] seconds!")
+				return
+			if(!line_of_sight(xeno_owner, target, WORLD_VIEW_NUM))
+				xeno_owner.balloon_alert(xeno_owner, "No line of sight with [target]")
```
generally don't use names in balloon alerts because they can become huge and annoying. They should be as short as possible while conveying the useful information.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+/// Adds effects for when we make impact against something.
+/datum/action/ability/activable/xeno/landslide/proc/on_impact(datum/source, atom/A, speed)
+	SIGNAL_HANDLER
+	var/turf/T = get_turf(A)
```
going to stop commenting on every instance one 1 letter vars but please fix it

lKiseki replied:
Changed all one letter vars to be more detailed. If I missed any, let me know.

Lumipharon reviewed (changes requested):
First pass. Lots more but a whole bunch of this stuff is just repeated minor issues for you to sort out

lKiseki commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+		COMSIG_OBJ_TRY_ALLOW_THROUGH = PROC_REF(can_climb_over),
+		COMSIG_FIND_FOOTSTEP_SOUND = TYPE_PROC_REF(/atom/movable, footstep_override),
+		COMSIG_TURF_CHECK_COVERED = TYPE_PROC_REF(/atom/movable, turf_cover_check),
+	)
```
These are added and removed multiple times depending on stuff done to the Earth Pillar, and having them as a var was convenient, as opposed to just setting them up on `Initialize()`.

Lumipharon replied:
just put it in a dedicated proc with the static list there

lKiseki replied:
Done.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/behemoth.dm:
```diff
@@ -1,55 +1,25 @@
+		return
+	med_hud_set_health()
+	update_wounds()
 	return TRUE
```
doesn't that simply not allow any updates, as opposed to removing wounds entirely?

lKiseki replied:
Done in #18533.

lKiseki commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
-	layer = BELOW_MOB_LAYER
-	color = COLOR_ORANGE
+/// Whenever the current holder moves, this object will follow.
+/obj/structure/xeno/earth_pillar/proc/on_move(datum/source, atom/old_loc, movement_dir, forced, list/old_locs)
```
Changed some proc names to be clearer. If I missed any, let me know.

lKiseki commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1217 @@
+/obj/structure/xeno/earth_pillar/proc/update_visuals(datum/source)
+	SIGNAL_HANDLER
+	update_appearance()
+	var/image/holder = hud_list[HEALTH_HUD_XENO] // TO DO: why is this HUD shit not working? send help
```
[Lumi pointed out the mistake](https://discord.com/channels/498569646014857217/500801682058379265/1448571111821279372). This is solved.

lKiseki committed:
behemoth rework: 1st review compliance

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -117,19 +110,31 @@
 	return TRUE
 
+/// Enables or disables connections used to check for climbing and other stuff.
+/obj/structure/xeno/earth_pillar/proc/handle_connections(toggle)
```
Its not a big deal but 'toggle' as an arg is not really clear. What you're actually doing is turning it ON or OFF. Toggle implies TRUE means its swapping state regardless of what it currently is, which its not doing.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -440,18 +447,18 @@
+		step_away(atom_checked, target, 1, 1) // Displaced by a pillar spawning underneath.
+	var/obj/structure/xeno/earth_pillar/new_pillar = new(target, xeno_owner, creation_limit > initial(creation_limit) ? FOUNDATIONS_HEALTH_REDUCTION : FALSE)
+	active_pillars += new_pillar
+	RegisterSignals(new_pillar, list(COMSIG_QDELETING), PROC_REF(pillar_deleted))
```
why signals when its one sig?

Lumipharon reviewed (changes requested):
some review on the updates

lKiseki commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -117,19 +110,31 @@
+	)
+	if(!toggle)
+		RemoveElement(/datum/element/connect_loc)
+		return
```
Donezo.

lKiseki commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/mutations_behemoth.dm:
```diff
@@ -1,12 +1,8 @@
-// I need this to load first because I wanna use the defines here, which is why it has an underscore in the file name.
-// Also, I didn't place much care into these because mutations were soft removed at the time.
+// I didn't place much care into these because mutations were soft removed at the time.
```
I did fix it!

lKiseki committed:
behemoth rework: missing stuff from 1st review

TiviPlus commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1474 +2,1224 @@
+// ***************************************
+// *********** Earth Pillar
+// ***************************************
+#define isearthpillar(A) (istype(A, /obj/structure/xeno/earth_pillar)) // So that we don't need to repeat the type path.
```
this should be in the file with the other isx defines

UltraBigRat commented:
Also the rock being an item seems a bit finnicky since it makes you need to swap hands, and you cant hit stuff with it or throw it anyways without using the ability.

Past those issues though do like the new direction for behemoth!

UltraBigRat commented:
Played it and this is a pretty awesome rework. My only real issue was how long the cooldown on getting a new rock is, but its a pretty fair balance decision still with how strong it is here.

TiviPlus commented:
fix conflict for metge

lKiseki commented:
> 1. I think hitting someone directly with a pillar should do some damage
>    Very weird to nail someone in the forehead with a huge ass boulder and all it does is knock them down (under the rock)

I'll discuss this with Lumi to see what they think.

> 2. Landsliding into a vehicle allows you to slide under/pass it
>    Considering how fucking awkward it is to engage a vehicle sometimes especially with a telegraphed move like Landslide
>    imagine you are parallel (or become parallel) to the hull and it drives into you and you decide to ram into it with Landslide (and hopefully hop on)
>    or you slide underneath and you get pushed backwards (into a marine mob) uhhhhhhh uhhhhhhhhhhh uhhhhhh fuck that

This is an issue with most forms of throws (such as Runner and Hunter's pounces, Defender's charge, and so on). The xeno will just go under the tank, which can result in awkward scenarios like these. I'll see what I can do to fix it for Behemoth, at least.

> 3. Geocrush can't be used against vehicles, don't know about structures (though it'd make it more like warrior punch x2 so I have no idea if you want it to do that)

More walance stuff, so I'll also discuss it with Lumi.

> 4. It's still 2025 here

not anymore it isn't >:)

lKiseki committed:
behemoth rework: bugfixes and review compliance

lKiseki committed:
Merge branch 'master' of https://github.com/tgstation/TerraGov-Marine-Corps into beherework

lKiseki committed:
behemoth rework: stat changes, geocrush now affects objs

lKiseki commented:
## List of changes since last test merge
### Stat changes
- Melee damage reduced from 28 to 26.
- Armor penetration increased from 10 to 15.
### Earth Riser
- Hitting a target directly with a rock (as in, having it land on top of them) will now deal brute damage equal to the Behemoth's slash damage.
### Earth Pillars
- Reduced their armor against melee attacks and bullets.
  - _Before:_ 60 melee and 50 bullet.
  - _After:_ 40 melee and 40 bullet.
- Health bar is now visible to everyone, not just xenos. Besides being good for clarity, it's also a way for me to compensate for the lack of damage representation on the icons, since I couldn't get those.
### Geocrush
- Increased the percentage of stamina damage dealt by this ability, from 50% of the Behemoth's slash damage to 100% of it.
- This ability can now hit objects of all kinds (including vehicles like APCs and tanks), not just living mobs, dealing an amount of damage equal to the Behemoth's slash damage. Against objects specifically, this is multiplied by 4.
### Bug fixes
- Fixed a bug where an **Earth Pillar's** health bar wouldn't update in certain circumstances.
- Fixed a bug where **Geocrush** could hit allied xenos.
- Removed a leftover message for admins from debugging (hi Titan).
### Code stuff
- Moved `isearthpillar` to `is_helpers.dm`, together with other similar procs.
- Added various new comments, and edited some existing ones, to better explain what I'm doing throughout the file.
- Earth Pillars now use the climbing component instead of the var. 
- Removed some commented code regarding Earth Pillar icons. I couldn't get these, but the health bar being visible to everyone should make up for it until then.
- Adjusted **Geocrush's** target acquisition, and relevant checks, to account for the changes specified above.
- Adjusted button code to be compatible with **Geocrush's** intended functionality.

lKiseki commented:
# HI MAINTAINERS PLEASE TEST MERGE THIS DO NOT MERGE IT I MADE BIG CHANGES THAT NEED TESTING
thank you

lKiseki committed:
behemoth rework: remove debugging messages, exclude cables from geocrush

Lumipharon commented on code/__DEFINES/mobs.dm:
```diff
@@ -861,6 +861,13 @@ GLOBAL_LIST_INIT(layers_to_offset, list(
+#define FOUNDATIONS_HEALTH_REDUCTION 0.4 // Health reduction applied to Earth Pillars as a percentage
+#define EARTH_MIGHT_ADDITIONAL_DAMAGE 1.3 // Additional damage provided by Earth Might as a percentage
+#define EARTH_MIGHT_PILLAR_DAMAGE 0.5 // Amount of damage done to Earth Pillars as a percentage
+#define GUIDED_CLAIM_ADDITIONAL_RANGE 2 // Additional range given to Seize
```
not a big deal but you can autodoc defines, so there's really no reason to not do so, if you're already adding comments

lKiseki replied:
Done in next commit.

Lumipharon commented on code/datums/personal_statistics.dm:
```diff
@@ -104,6 +104,15 @@ GLOBAL_LIST_EMPTY(personal_statistics_list)
 	var/acid_maw_uses = 0
 	var/acid_jaw_uses = 0
 
+	// Behemoth stats
+	var/seize_uses = 0 // Amount of times we used Seize successfully.
+	var/earth_pillars_created = 0 // Amount of Earth Pillars we created.
+	var/earth_pillars_thrown = 0 // Amount of Earth Pillars we threw.
+	var/earth_pillar_repairs = 0 // Amount of integrity we repaired on Earth Pillars.
+	var/landslide_damage = 0 // Amount of damage dealt with Landslide.
+	var/geocrush_damage = 0 // Amount of damage dealt with Geocrush.
+	var/primal_wrath_healing = 0 // Amount of health we recovered with Primal Wrath.
+
```
as above, why add comments instead of just autodoc

lKiseki replied:
Done in next commit.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/abilities.dm:
```diff
@@ -1530,3 +1530,27 @@ GLOBAL_LIST_INIT(pattern_images_list, list(
+// ***************************************
+// *********** Xeno Warnings
+// ***************************************
+/obj/effect/temp_visual/xeno_warning
```
this implies its generic, but is it actually generic or specific to bemo?

lKiseki replied:
This is generic and not exclusive to Behemoth. It's a way for any xenomorph to have warnings on their abilities (and something I intend to use in the future, if I ever do any significant changes to other castes).

Lumipharon replied:
the reason I ask is it should be clear what they are warning ABOUT. i.e. generic is bad unless you can visually differentiate between bemo warning vs oppressor warning etc etc

lKiseki replied:
You do get an option to choose a color for the xeno warnings (it's done when the effect initializes, if I recall correctly), so that's how you can differentiate them.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+	// Using grab intent on a rock will let us grab it.
+	if(xeno_attacker.a_intent == INTENT_GRAB)
+		when_grabbed(xeno_attacker)
+		return
```
don't need to update it, but FYI for readability its better to put this sort of early return before the help intent section, since the help intent part doesn't (always) early return while this does.

lKiseki replied:
Done in next commit.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+/// Cleans up after this object is thrown.
+/obj/structure/xeno/earth_pillar/proc/post_throw(datum/source)
+	SIGNAL_HANDLER
+	setup_connections()
```
why's this running again? the whole point of connect_loc is that it stays connected to it's loc as it moves

lKiseki replied:
The idea here is to failproof the object from any unwanted interactions when it's thrown. To do this, I remove the `connect_loc` element before it is thrown, and after the throw concludes, I set it up again. If you think it is unnecessary, I can just keep the element instead of removing it like this.

Lumipharon replied:
thats fine then

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
 		return
-	return ..()
+	// Shouldn't be able to use this ability if you can't move.
+	if(!xeno_owner.canmove)
+		return FALSE
```
just a side note, does being immobilised or stunned (as in the status effect) actually set this?

lKiseki replied:
No. The only time this is set to something else is when using the rappel system, but xenos can't use it anyway. I intended to set it on Primal Wrath, but this ended up being unnecessary, so I'll just remove checks like these since they're irrelevant at this point.

lKiseki replied:
I lied, actually -- there's a proc called `set_canmove` which I use in some other parts of Behemoth's code. It's also used with the **floored** and **immobile** status effects (or traits; I'm not sure of what they are anymore).

Lumipharon replied:
floored and immobile is what stuns actually are in the end

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
-		playsound(object_turf, 'sound/effects/meteorimpact.ogg', 30, TRUE)
-		new /obj/effect/temp_visual/behemoth/landslide/hit(object_turf)
-		vending_target.tip_over()
+	if(!line_of_sight(xeno_owner, target, WORLD_VIEW_NUM))
```
do you really want to be drawing LOS for the entire map instead of just view range?

lKiseki replied:
I thought this was view range, no? `WORLD_VIEW_NUM` is 7, which is the standard view range as far as I'm aware.

Lumipharon replied:
completely misread the var, sorry. is 7 fine or are you actually checking real view range? i.e. widescreen?

lKiseki replied:
I was checking real view range, but I'm not sure how I feel about accounting for widescreen view. Maybe seven is fine regardless?

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+/// Any atoms on top will be dragged along when this one is thrown by this ability.
+/atom/movable/proc/geocrush_drag(datum/source, atom/old_loc, movement_dir, forced, list/old_locs)
+	for(var/atom/movable/movable_atom AS in old_loc)
+		movable_atom.forceMove(loc)
```
why is this force moving EVERYTHING?

lKiseki replied:
I've decided to use Warrior's `punch_act` proc to resolve effects against most objects and structures, while only leaving exceptions for xeno structures and living mobs. This should be much easier, especially since Warrior's Punch ability is a good enough precedent for what I'm trying to accomplish here. This comment should also apply to the rest of your review down below, so I'm just gonna put "ditto" on those.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+		if(!allcut)
+			wires.cut_all()
+	update_appearance()
+	return ..()
```
I'm not paying attention to walance right now but ranged apc disable is aylmao

lKiseki replied:
Ditto, see above comment.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+
+/obj/machinery/computer/geocrush_act(...)
+	set_disabled()
+	return ..()
```
is this disabling indestructable computers?

lKiseki replied:
Ditto, see above comment.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+
+/obj/machinery/keycard_auth/geocrush_act(...)
+	qdel(src)
+	return TRUE
```
a lot of these in general seem highly questionable and are letting xenos interact with things (or outright delete them) that they shouldn't normally be able to.

lKiseki replied:
Ditto, see above comment.

Lumipharon commented on code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm:
```diff
@@ -2,1473 +2,1643 @@
+	qdel(src)
+	return TRUE
+
+/obj/structure/prop/geocrush_impact(atom/causer)
```
honestly this entire list is more questionable the more I look at it. I get having this as an atom level proc and childing out the effects, but this is an insane list with lots of really wonky and snowflake interactions.

lKiseki replied:
Ditto, see above comment.

Lumipharon reviewed (changes requested):
Most of this is OK at a quick review, but some of these geocrush interactions are pretty cursed.

lKiseki committed:
behemoth rework: more autodoc, stat changes, geocrush is not cursed
